js/src/util.js

```diff
@@ -119,18 +136,54 @@ const Util = (() => {
         return {}
       }
 
-      const attributes = {}
-      for (let i = 0; i < element.attributes.length; i++) {
-        const attribute = element.attributes[i]
-        if (attribute.nodeName.indexOf('data-') !== -1) {
-          // remove 'data-' part of the attribute name
-          const attributeName = attribute.nodeName.substring('data-'.length)
-          attributes[attributeName] = attribute.nodeValue
+      let attributes
+      if (Object.getOwnPropertyDescriptor(HTMLElement.prototype, 'dataset')) {
+        attributes = this.extend({}, element.dataset)
+      } else {
+        attributes = {}
+        for (let i = 0; i < element.attributes.length; i++) {
+          const attribute = element.attributes[i]
+          if (attribute.nodeName.indexOf('data-') !== -1) {
+            // remove 'data-' part of the attribute name
+            const attributeName = attribute.nodeName.substring('data-'.length).replace(/-./g, (str) => str.charAt(1).toUpperCase())
+            attributes[attributeName] = attribute.nodeValue
+          }
+        }
+      }
+
+      for (const key in attributes) {
+        if (!Object.prototype.hasOwnProperty.call(attributes, key)) {
+          continue
         }
+
+        attributes[key] = normalizeData(attributes[key])
       }
+
       return attributes
     },
 
+    getDataAttribute(element, key) {
+      return normalizeData(element.getAttribute(`data-${key.replace(/[A-Z]/g, (chr) => `-${chr.toLowerCase()}`)}`))
+    },
+
+    setDataAttribute(element, key, value) {
+      const $ = this.jQuery
+      if (typeof $ !== 'undefined') {
+        $(element).data(key, value)
+      }
+
+      element.setAttribute(`data-${key.replace(/[A-Z]/g, (chr) => `-${chr.toLowerCase()}`)}`, value)
+    },
+
+    removeDataAttribute(element, key) {
```

Comment:
`setDataAttribute` and `removeDataAttribute` should be in our DOM Manipulator here : https://github.com/twbs/bootstrap/blob/v4-without-jquery/js/src/dom/manipulator.js